source/Calamari.Shared/Integration/Packages/Download/S3PackageDownloader.cs

```diff
@@ -87,35 +91,35 @@ string BuildFileName(string prefix, string version, string extension)
 
                     using (var s3Client = GetS3Client(feedUsername, feedPassword, region))
                     {
-                        bool fileExists = false;
-                        string fileName = string.Empty;
-                        string knownExtension = string.Empty;
-                        for (int i = 0; i < knownFileExtensions.Length && !fileExists; i++)
+                        string? foundFilePath = null;
+                        for (int i = 0; i < knownFileExtensions.Length && foundFilePath == null; i++)
                         {
-                            fileName = BuildFileName(prefix, version.ToString(), knownFileExtensions[i]);
-                            fileExists = FileExistsInBucket(s3Client, bucketName, fileName, CancellationToken.None)
+                            var fileName = BuildFileName(prefix, version.ToString(), knownFileExtensions[i]);
+                            foundFilePath = FindSingleFileInTheBucket(s3Client, bucketName, fileName, CancellationToken.None)
 #if NET40
                                 .Result;
 #else
                                          .GetAwaiter()
                                          .GetResult();
 #endif
-                            if (fileExists)
-                            {
-                                knownExtension = knownFileExtensions[i];
-                            }
                         }
 
-                        if (!fileExists)
+                        var fullFileName = "";
+                        if (foundFilePath != null)
+                        {
+                            fullFileName = foundFilePath;
+                        }
+
+                        if (fullFileName.IsNullOrEmpty())
                             throw new Exception($"Unable to download package {packageId} {version}: file not found");
 
-                        var localDownloadName = Path.Combine(cacheDirectory, PackageName.ToCachedFileName(packageId, version, knownExtension));
+                        var localDownloadName = Path.Combine(cacheDirectory, PackageName.ToCachedFileName(packageId, version, Extension));
```

Comment:
Note this extension type is hardcoded to a zip file, we do this in Server, but this may not be valid to do in Calamari. 

Is this reasonable for Linux workers?
Does this work for all package types?

Reply:
Good question and I wish I knew the answer 😆 From my understanding feeds don't always need to extract the package file but just pass it through, where the extension could be anything. Looking at the code though this seems wrong. What is supposed to happen is we extract the archive as it's stored in S3 and rezip it. Which wouldn't work with the case where we are just passing the file through. This needs some manual testing.

Reply:
Looks like there were issues at least in the case of the double extension .tar.gz file types. I've updated this to iterate over the known file extensions, fall back to the file extension from the file itself (this doesn't handle double extensions hence it's usage as a fallback) and finally fall back to zip. I'm not sure if this is a potential issue in the Server implementation.